File: 113_DijkstraAlgorithm.py
"""
Problem: Dijkstra's Algorithm
Description: Find shortest paths from source vertex to all other vertices in a
             weighted graph with non-negative edge weights.

Approach: Use priority queue (min-heap). Greedily pick the unvisited vertex with
          smallest distance, update distances of its neighbors.
"""

from typing import List, Tuple
import heapq

def dijkstra(V: int, adj: List[List[Tuple[int, int]]], src: int) -> List[int]:
    dist = [float('inf')] * V
    pq = [(0, src)]
    dist[src] = 0
    
    while pq:
        d, u = heapq.heappop(pq)
        
        if d > dist[u]:
            continue
        
        for v, weight in adj[u]:
            if dist[u] + weight < dist[v]:
                dist[v] = dist[u] + weight
                heapq.heappush(pq, (dist[v], v))
    
    return dist

def main():
    V = 5
    adj = [[] for _ in range(V)]
    
    adj[0].append((1, 2))
    adj[0].append((2, 4))
    adj[1].append((2, 1))
    adj[1].append((3, 7))
    adj[2].append((4, 3))
    adj[3].append((4, 1))
    
    distances = dijkstra(V, adj, 0)
    
    print("Shortest distances from source 0:")
    for i in range(V):
        if distances[i] == float('inf'):
            print(f"{i}: INF")
        else:
            print(f"{i}: {distances[i]}")

if __name__ == "__main__":
    main()

"""
Time Complexity: O((V + E) log V)
Space Complexity: O(V)
"""

File: 113_DijkstraAlgorithm.cpp
/*
Problem: Dijkstra's Algorithm
Description: Find shortest paths from source vertex to all other vertices in a
             weighted graph with non-negative edge weights.

Approach: Use priority queue (min-heap). Greedily pick the unvisited vertex with
          smallest distance, update distances of its neighbors.
*/

#include <bits/stdc++.h>
using namespace std;

vector<int> dijkstra(int V, vector<vector<pair<int, int>>>& adj, int src) {
    vector<int> dist(V, INT_MAX);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    
    dist[src] = 0;
    pq.push({0, src});
    
    while (!pq.empty()) {
        int u = pq.top().second;
        int d = pq.top().first;
        pq.pop();
        
        if (d > dist[u]) continue;
        
        for (auto& [v, weight] : adj[u]) {
            if (dist[u] + weight < dist[v]) {
                dist[v] = dist[u] + weight;
                pq.push({dist[v], v});
            }
        }
    }
    
    return dist;
}

int main() {
    int V = 5;
    vector<vector<pair<int, int>>> adj(V);
    
    adj[0].push_back({1, 2});
    adj[0].push_back({2, 4});
    adj[1].push_back({2, 1});
    adj[1].push_back({3, 7});
    adj[2].push_back({4, 3});
    adj[3].push_back({4, 1});
    
    vector<int> distances = dijkstra(V, adj, 0);
    
    cout << "Shortest distances from source 0:" << endl;
    for (int i = 0; i < V; i++) {
        if (distances[i] == INT_MAX)
            cout << i << ": INF" << endl;
        else
            cout << i << ": " << distances[i] << endl;
    }
    
    return 0;
}

/*
Time Complexity: O((V + E) log V)
Space Complexity: O(V)
*/
